### PR TITLE
speed up `deepcopy` on pointer arrays

### PR DESCRIPTION
**Pull request by JeffBezanson**
Before:

```
julia> a = convert(Array{Any}, rand(1000,1000));

julia> @time deepcopy(a)
  1.101568 seconds (3.01 M allocations: 69.428 MB, 14.23% gc time)
```

after:

```
julia> a = convert(Array{Any}, rand(1000,1000));

julia> @time deepcopy(a)
  0.057694 seconds (17.21 k allocations: 8.424 MB)
```

### CODE REVIEW

File: base/deepcopy.jl [hunk 1]
```diff
@@ -23,12 +23,12 @@ function deepcopy_internal(x::SimpleVector, stackdict::ObjectIdDict)
...
 function deepcopy_internal(x::ANY, stackdict::ObjectIdDict)
-    if haskey(stackdict, x)
-        return stackdict[x]
-    end
     T = typeof(x)::DataType
     nf = nfields(T)
     (isbits(T) || nf == 0) && return x
+    if haskey(stackdict, x)
+        return stackdict[x]
+    end
```
Comment (StefanKarpinski) [inline base/deepcopy.jl]: How does moving this help?

**Review by vtjnash — APPROVED** [review]

File: base/deepcopy.jl (hunk 1 above)
Comment (JeffBezanson) [inline base/deepcopy.jl]: It's much less important than the rest of the change, but checking for `isbits` is faster than doing a dictionary lookup, plus isbits values are never added to the dictionary, so this lets us skip the lookup entirely in those cases.

**Review by StefanKarpinski — APPROVED** [review]

**Comment by JeffBezanson** [comment]
Just a heads up that this may not apply cleanly to 0.5 due to the change from `isdefined` to `isassigned` on arrays. Simple fix tho.

**Comment by tkelman** [comment]
`jl_array_isdefined` is not exported from libjulia on release-0.5. `jl_field_isdefined` is though.

**Comment by yuyichao** [comment]
This is using `jl_array_isassigned`?

**Comment by tkelman** [comment]
> Just a heads up that this may not apply cleanly to 0.5 due to the change from `isdefined` to `isassigned`

asking for the purposes of backporting

**Comment by yuyichao** [comment]
But none of the two functions you mentioned is used by this PR?

**Comment by tkelman** [comment]
ah right, nevermind, misread what Jeff meant

**Pull request merged**